modules/src/applications/transfer/msgs/transfer.rs

```diff
@@ -63,8 +63,27 @@ impl TryFrom<RawMsgTransfer> for MsgTransfer {
         let timeout_timestamp = Timestamp::from_nanoseconds(raw_msg.timeout_timestamp)
             .map_err(|_| Error::invalid_packet_timeout_timestamp(raw_msg.timeout_timestamp))?;
 
+        // `RawPacket.timeout_height` is treated differently because
+        //
+        // `RawHeight.timeout_height == {revision_number: 0, revision_height = 0}`
+        //
+        // is legal and meaningful, even though the Tendermint spec rejects this
+        // height as invalid. Thus, it must be parsed specially, where this
+        // special case means "no timeout".
+        //
+        // Note: it is important for `revision_number` to also be `0`, otherwise
+        // packet commitment proofs will be incorrect (see proof construction in
+        // `ChannelReader::packet_commitment()`). Note also that ibc-go conforms
+        // to this.
         let timeout_height: Option<Height> = raw_msg
             .timeout_height
+            .and_then(|raw_height| {
+                if raw_height.revision_number == 0 && raw_height.revision_height == 0 {
+                    None
+                } else {
+                    Some(raw_height)
+                }
+            })
```

Comment:
To clean this up a bit, can we perhaps add a `From<RawHeight>` impl for `Option<Height>`?

Or perhaps just use `Height::new(raw_height.revision_number, raw_height.revision_height).ok()`.

I think the former instance might be useful but perhaps the latter option is cleaner, what do you think?

Reply:
`TimeoutHeight` def: [5660a99](https://github.com/informalsystems/ibc-rs/pull/2346/commits/5660a999ac192a0fc29d1de1797161ac212782c3)
Propagate changes: [69234fb](https://github.com/informalsystems/ibc-rs/pull/2346/commits/69234fbe7d3ee615f5dcb3dbca37203ba4aec30d)